src/Lean/Data/LParsec.lean

```diff
@@ -0,0 +1,138 @@
+
+/-
+Copyright (c) 2021 Microsoft Corporation. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Dany Fabian
+-/
+
+namespace Lean
+
+namespace LParsec
+
+inductive ParseResult (α : Type) where
+  | success (pos : String.Iterator) (res : α)
+  | error (pos : String.Iterator) (err : String)
+  deriving Repr
+
+open ParseResult
+
+def Parser (α : Type) : Type := String.Iterator → ParseResult α
+
+instance (α : Type) : Inhabited (Parser α) :=
+  ⟨λ it => error it ""⟩
+
+@[inline]
+protected def pure (a : α) : Parser α := λ it => 
+ success it a 
+ 
+@[inline]
+def bind {α β : Type} (f : Parser α) (g : α → Parser β) : Parser β := λ it =>
+  match f it with
+  | success rem a => g a rem
+  | error pos msg => error pos msg
+  
+@[inline]
+instance : Monad Parser :=
+  { pure := LParsec.pure, bind }
+  
+@[inline]
+def fail (msg : String) : Parser α := fun it =>
+  error it msg
+
+@[inline]
+def orElse (p q : Parser α) : Parser α := fun it =>
+  match p it with
+  | success rem a => success rem a
+  | error _ _ => q it
+
+@[inline]
+def attempt (p : Parser α) : Parser α := λ it =>
+  match p it with
+  | success rem res => success rem res
+  | error _ err => error it err
+
+@[inline]
```

Comment:
You shouldn't usually have to mark instances as `[inline]`, they are inlined automatically when a class field is accessed.